nexus/src/app/background/lookup_region_port.rs

```diff
@@ -0,0 +1,149 @@
+use nexus_types::internal_api::background::LookupRegionPortStatus;
+use serde_json::json;
+use std::net::SocketAddrV6;
+use std::sync::Arc;
+use uuid::Uuid;
+
+pub struct LookupRegionPort {
+    datastore: Arc<DataStore>,
+}
+
+impl LookupRegionPort {
+    pub fn new(datastore: Arc<DataStore>) -> Self {
+        LookupRegionPort { datastore }
+    }
+}
+
+async fn get_region_from_agent(
+    agent_address: &SocketAddrV6,
+    region_id: Uuid,
+) -> Result<Region> {
+    let url = format!("http://{}", agent_address);
+    let client = CrucibleAgentClient::new(&url);
+
+    let result = client.region_get(&RegionId(region_id.to_string())).await?;
+
+    Ok(result.into_inner())
+}
+
+impl BackgroundTask for LookupRegionPort {
+    fn activate<'a>(
+        &'a mut self,
+        opctx: &'a OpContext,
+    ) -> BoxFuture<'a, serde_json::Value> {
+        async {
+            let log = &opctx.log;
+            info!(&log, "lookup region port task started");
+
+            let mut status = LookupRegionPortStatus::default();
+
+            let regions_missing_ports =
+                match self.datastore.regions_missing_ports(opctx).await {
+                    Ok(regions) => regions,
+
+                    Err(e) => {
+                        let s = format!(
+                            "could not find regions missing ports: {e}"
+                        );
+
+                        error!(log, "{s}");
+                        status.errors.push(s);
+
+                        return json!(status);
+                    }
+                };
+
+            for region in regions_missing_ports {
+                let dataset_id = region.dataset_id();
+
+                let dataset = match self.datastore.dataset_get(dataset_id).await
+                {
+                    Ok(dataset) => dataset,
+
+                    Err(e) => {
+                        let s =
+                            format!("could not get dataset {dataset_id}: {e}");
+
+                        error!(log, "{s}");
+                        status.errors.push(s);
+
+                        continue;
+                    }
+                };
+
+                let returned_region = match get_region_from_agent(
+                    &dataset.address(),
+                    region.id(),
+                )
+                .await
+                {
+                    Ok(returned_region) => returned_region,
+
+                    Err(e) => {
+                        let s = format!(
+                            "could not get region {} from agent: {e}",
+                            region.id(),
+                        );
+
+                        error!(log, "{s}");
+                        status.errors.push(s);
+
+                        continue;
+                    }
+                };
+
+                match self
+                    .datastore
+                    .region_set_port(region.id(), returned_region.port_number)
+                    .await
```

Comment:
So it's possible that multiple distinct nexus instances call this function concurrently.

I'm assuming this is safe because the function is deterministic, and we're "pulling" the port number from the agent, rather than doing an allocation, right?

Reply:
Correct yeah.